### PR TITLE
Add `DeepStateModel` 

### PR DESCRIPTION
**Pull request by ostreech1997**
## Before submitting (must do checklist)

- [x] Did you read the [contribution guide](https://github.com/tinkoff-ai/etna/blob/master/CONTRIBUTING.md)?
- [x] Did you update the docs? We use Numpy format for all the methods and classes.
- [x] Did you write any new necessary tests?
- [x] Did you update the [CHANGELOG](https://github.com/tinkoff-ai/etna/blob/master/CHANGELOG.md)?

## Proposed Changes

Implementation of the Deep State Space Model

### CODE REVIEW

File: etna/models/nn/deepstate/state_space_model.py [hunk 1]
```diff
@@ -0,0 +1,445 @@
...
+from torch import Tensor
+from torch.nn.functional import one_hot
+
+from etna.core import BaseMixin
+
+
+class SSM(ABC, BaseMixin):
+    """Base class for State Space Model.
+
+    The system dynamics is described with the following equations:
```
Comment (martins0n) [inline etna/models/nn/deepstate/state_space_model.py:20]: image

It seems not rendering correctly

**Review by martins0n — CHANGES_REQUESTED** [review]
Can we add examples to neural network notebook?
And we should add some information about choosing right family of SSM

**Comment by martins0n** [comment]
And yet another thing - look at test inference folder - it would be great to add test cases for deepstate there

File: etna/models/nn/deepstate/state_space_model.py [hunk 2]
```diff
@@ -0,0 +1,445 @@
...
+        datetime_index:
+            Tensor with the index values.
+            Values should be from 0 to seasonal period.
+
+        Returns
+        -------
+        :
+            Transition coefficient matrix.
+        """
+        transition_coeff = torch.eye(self.latent_dim())
```
Comment (alex-hse-repository) [inline etna/models/nn/deepstate/state_space_model.py:159]: Don't we need to set the device here as well -- I guess this should solve the problems in deepstate

File: etna/models/nn/deepstate/state_space_model.py [hunk 3]
```diff
@@ -0,0 +1,445 @@
...
```
Comment (alex-hse-repository) [inline etna/models/nn/deepstate/state_space_model.py:222]: The same about device

File: etna/models/nn/deepstate/state_space_model.py [hunk 4]
```diff
@@ -0,0 +1,445 @@
...
+        datetime_index:
+            Tensor with the index values.
+            Values should be from 0 to seasonal period.
+
+        Returns
+        -------
+        :
+            Emission coefficient matrix.
+        """
+        emission_coeff = one_hot(datetime_index.squeeze(-1), num_classes=self.latent_dim())
```
Comment (alex-hse-repository) [inline etna/models/nn/deepstate/state_space_model.py:265]: The same about device

File: etna/models/nn/deepstate/state_space_model.py [hunk 5]
```diff
@@ -0,0 +1,445 @@
...
+        Returns
+        -------
+        :
+            Datetime index for State Space Model.
+        """
+        return np.array([self.timestamp_transform(timestamp) for timestamp in timestamps])
+
+
+class YearlySeasonalitySSM(SeasonalitySSM):
+    """Class for Weekly Seasonality State Space Model."""
```
Comment (alex-hse-repository) [inline etna/models/nn/deepstate/state_space_model.py:285]: Yearly

File: etna/models/nn/deepstate/linear_dynamic_system.py [hunk 6]
```diff
@@ -0,0 +1,226 @@
...
+        latent_dim: int,
+    ):
+        """Create instance of LDS.
+
+        Parameters
+        ----------
+        emission_coeff:
+            Emission coefficient matrix with shape (batch_size, seq_length, latent_dim).
+        transition_coeff:
+            Transition coefficient matrix with shape (latent_dim, latent_dim).
```
Comment (alex-hse-repository) [inline etna/models/nn/deepstate/linear_dynamic_system.py:33]: Can you make sure that transition_coeff is always time independent

File: etna/models/nn/deepstate/linear_dynamic_system.py [hunk 7]
```diff
@@ -0,0 +1,226 @@
...
+        self.innovation_coeff = innovation_coeff
+        self.noise_std = noise_std
+        self.prior_mean = prior_mean
+        self.prior_cov = prior_cov
+        self.offset = offset
+        self.seq_length = seq_length
+        self.latent_dim = latent_dim
+
+        self.batch_size = self.prior_mean.shape[0]
+        self._eye = torch.eye(self.latent_dim).type_as(noise_std)
```
Comment (alex-hse-repository) [inline etna/models/nn/deepstate/linear_dynamic_system.py:60]: `to`

File: etna/models/nn/deepstate/linear_dynamic_system.py [hunk 8]
```diff
@@ -0,0 +1,226 @@
...
+        # H * mu (batch_size, 1)
+        target_mean = (emission_coeff.permute(0, 2, 1) @ prior_mean.unsqueeze(-1)).squeeze(-1)
+        # v (batch_size, 1)
+        residual = target - target_mean - offset
+        # R (batch_size, 1, 1)
+        noise_cov = torch.diag_embed(noise_std * noise_std)
+        # F (batch_size, 1, 1)
+        target_cov = emission_coeff.permute(0, 2, 1) @ prior_cov @ emission_coeff + noise_cov
+        # K (batch_size, latent_dim)
+        kalman_gain = (prior_cov @ emission_coeff @ torch.inverse(target_cov)).squeeze(-1)
```
Comment (alex-hse-repository) [inline etna/models/nn/deepstate/linear_dynamic_system.py:111]: Actually `inverse` is a dummy implementation -- it is not the best practice to implement it this way. May be you can search for better implementation(see gluons one)

File: etna/models/nn/deepstate/deepstate.py [hunk 9]
```diff
@@ -0,0 +1,317 @@
...
+        self.num_layers = num_layers
+        self.n_samples = n_samples
+        self.lr = lr
+        self.optimizer_params = {} if optimizer_params is None else optimizer_params
+        self.latent_dim = self.ssm.latent_dim()
+
+        self.RNN = nn.LSTM(
+            num_layers=self.num_layers, hidden_size=self.latent_dim, input_size=self.input_size, batch_first=True
+        )
+        self.projectors = nn.ModuleDict(
```
Comment (alex-hse-repository) [inline etna/models/nn/deepstate/deepstate.py:68]: In the paper there is also the third type of transformations for the bounded parameters -- may be we can create the separate issue to add this option + prediction intervals

File: etna/models/nn/deepstate/deepstate.py [hunk 10]
```diff
@@ -0,0 +1,317 @@
...
+        datetime_index = batch["datetime_index"].permute(1, 0, 2)[
+            :, :, :seq_length
+        ]  # (num_models, batch_size, seq_length)
+
+        output, (_, _) = self.RNN(encoder_real)  # (batch_size, seq_length, latent_dim)
+        prior_std = self.projectors["prior_std"](output[:, 0])
+
+        lds = LDS(
+            emission_coeff=self.ssm.emission_coeff(datetime_index),
+            transition_coeff=self.ssm.transition_coeff(datetime_index).type_as(targets),
```
Comment (alex-hse-repository) [inline etna/models/nn/deepstate/deepstate.py:107]: Don't actually need type_as if we fix it in SSM

File: etna/models/nn/deepstate/deepstate.py [hunk 11]
```diff
@@ -0,0 +1,317 @@
...
+        ]  # (num_models, batch_size, seq_length)
+        datetime_index_test = x["datetime_index"].permute(1, 0, 2)[
+            :, :, seq_length:
+        ]  # (num_models, batch_size, horizon)
+
+        output, (h_n, c_n) = self.RNN(encoder_real)  # (batch_size, seq_length, latent_dim)
+        prior_std = self.projectors["prior_std"](output[:, 0])
+        lds = LDS(
+            emission_coeff=self.ssm.emission_coeff(datetime_index_train),
+            transition_coeff=self.ssm.transition_coeff(datetime_index_train).type_as(targets),
```
Comment (alex-hse-repository) [inline etna/models/nn/deepstate/deepstate.py:148]: The same here

File: etna/models/nn/deepstate/deepstate.py [hunk 12]
```diff
@@ -0,0 +1,317 @@
...
+            seq_length=seq_length,
+            latent_dim=self.latent_dim,
+        )
+        _, prior_mean, prior_cov = lds.kalman_filter(targets=targets)
+
+        output, (_, _) = self.RNN(decoder_real, (h_n, c_n))  # (batch_size, horizon, latent_dim)
+        horizon = output.shape[1]
+        lds = LDS(
+            emission_coeff=self.ssm.emission_coeff(datetime_index_test),
+            transition_coeff=self.ssm.transition_coeff(datetime_index_test).type_as(targets),
```
Comment (alex-hse-repository) [inline etna/models/nn/deepstate/deepstate.py:163]: The same here

**Review by alex-hse-repository — CHANGES_REQUESTED** [review]

File: etna/models/nn/deepstate/state_space_model.py (hunk 4 above)
Comment (ostreech1997) [reply etna/models/nn/deepstate/state_space_model.py:265]: Device is defined by datetime_index here. I think, we don't have to change code.

File: etna/models/nn/deepstate/linear_dynamic_system.py (hunk 8 above)
Comment (ostreech1997) [reply etna/models/nn/deepstate/linear_dynamic_system.py:111]: `target_cov` is (batch_size, 1, 1). We can replace `torch.inverse(target_cov)` with `(1 / target_cov)`.

**Review by martins0n — DISMISSED** [review]

**Review by alex-hse-repository — APPROVED** [review]

**Review by alex-hse-repository — APPROVED** [review]

**Review by alex-hse-repository — APPROVED** [review]

**Pull request merged**